ydb/core/blobstorage/vdisk/balance/sender.cpp

```diff
@@ -0,0 +1,201 @@
+#include "defs.h"
+#include "utils.h"
+
+#include <ydb/core/blobstorage/vdisk/common/vdisk_queues.h>
+#include <ydb/core/blobstorage/base/vdisk_sync_common.h>
+
+
+namespace NKikimr {
+namespace NBalancing {
+namespace {
+    class TReader {
+    private:
+        const size_t BatchSize;
+        TPDiskCtxPtr PDiskCtx;
+        TQueue<TPartInfo> Parts;
+        TReplQuoter::TPtr Quoter;
+        const TBlobStorageGroupType GType;
+
+        TVector<TPart> Result;
+        ui32 Responses;
+        ui32 ExpectedResponses;
+    public:
+        enum EReaderState {
+            WAITING_PDISK_RESPONSES,
+            FINISHED,
+        };
+
+        TReader(size_t batchSize, TPDiskCtxPtr pDiskCtx, TQueue<TPartInfo> parts, TReplQuoter::TPtr replPDiskReadQuoter, TBlobStorageGroupType gType)
+            : BatchSize(batchSize)
+            , PDiskCtx(pDiskCtx)
+            , Parts(std::move(parts))
+            , Quoter(replPDiskReadQuoter)
+            , GType(gType)
+            , Result(Reserve(BatchSize))
+            , Responses(0)
+            , ExpectedResponses(0)
+        {}
+
+        EReaderState DoJobQuant(const TActorContext &ctx) {
+            if (ExpectedResponses != 0) {
+                return WAITING_PDISK_RESPONSES;
+            }
+            if (Parts.empty()) {
+                return FINISHED;
+            }
+            Result.resize(Min(Parts.size(), BatchSize));
+            ExpectedResponses = 0;
+            for (ui64 i = 0; i < BatchSize && !Parts.empty(); ++i) {
+                auto item = Parts.front();
+                Parts.pop();
+                Result[i] = TPart{
+                    .Key=item.Key,
+                    .Ingress=item.Ingress
+                };
+                if (std::holds_alternative<TRope>(item.PartData)) {
```

Comment:
std::visit() suits this case better, I suppose. You may also use TOverloaded{} to distinguish between types.